### PR TITLE
Handle instrument disposal

### PR DESCRIPTION
**Pull request by cijothomas**
Towards #2425 
This does not reclaim the position occupied by Metric in the Metric[], so the current hardcoded limit of 1000 remains even after disposing Meters (thereby instruments from it). This can be a follow up.

This will face similar challenges as https://github.com/open-telemetry/opentelemetry-dotnet/pull/2466 when reclaiming is incorporated, but likely a lot easier to fix as Instrument creation/disposal is not a hot path (instruments are typically created once and used throughout), and can afford to protect with plain lock statements.

This also does not re-allow the metric stream name even after the original instrument is disposed. There are some spec updates likely coming in this area, and this restriction might be going away. Hence not touching it in *this* PR.

### CODE REVIEW

File: docs/metrics/customizing-the-sdk/Program.cs [hunk 1]
```diff
@@ -60,7 +60,7 @@ public static void Main(string[] args)
             // turn off the above default. i.e any
             // instrument which does not match any views
             // gets dropped.
-            // .AddView(instrumentName: "*", new MetricStreamConfiguration() { Aggregation = Aggregation.Drop })
+            // .AddView(instrumentName: "*", MetricStreamConfiguration.Drop)
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/Program.cs:63]: unrelated minor fix

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 2]
```diff
@@ -135,7 +137,7 @@ internal sealed class MeterProviderSdk : MeterProvider
                         // which will apply defaults.
                         // Users can turn off this default
                         // by adding a view like below as the last view.
-                        // .AddView(instrumentName: "*", new MetricStreamConfiguration() { Aggregation = Aggregation.Drop })
+                        // .AddView(instrumentName: "*", MetricStreamConfiguration.Drop)
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:140]: unrelated minor fix

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 3]
```diff
@@ -412,12 +439,27 @@ internal Batch<Metric> Collect()
...
+                    int metricCountCurrentBatch = 0;
                     for (int i = 0; i < target; i++)
                     {
-                        this.metrics[i].SnapShot();
+                        var metric = this.metrics[i];
+                        if (metric != null)
+                        {
+                            if (metric.InstrumentDisposed)
+                            {
+                                metric.SnapShot();
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:450]: if the instrument was disposed *before* Collect, then only clear it in current Collect. Otherwise, it'll be cleaned up in the *next* collect only.

**Comment by reyang** [comment]
> This also does not re-allow the metric stream name even after the original instrument is disposed. There are some spec updates likely coming in this area, and this restriction might be going away. Hence not touching it in _this_ PR.

That's a good callout. It also has memory implications if someone keeps creating/disposing instruments.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:450]: What if the next Collect failed (e.g. timeout)?

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 3 above)
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:450]: nit: `metric.Snapshot()` is always called. Simplify code by only having `this.metrics[i] = null` in the `if` block.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:450]: The check `metric.InstrumentDisposed` must be done before snapshot all, so not sure how to rearrange the code here..

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:450]: good point. The provider currently has no way of knowing that (only readers know that).

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 4]
```diff
@@ -300,9 +305,31 @@ static Regex GetWildcardRegex(IEnumerable<string> collection)
...
+            metric.InstrumentDisposed = true;
+        }
+
         internal void MeasurementsCompleted(Instrument instrument, object state)
         {
-            Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} completed.");
+            var metrics = state as List<Metric>;
+            if (metrics == null)
+            {
+                // TODO: log
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:325]: test comment
